domain/consensus/finality_test.go

```diff
@@ -167,3 +172,248 @@ func TestFinality(t *testing.T) {
+			if err != nil {
+				t.Fatalf("TestBoundedMergeDepth: Failed to get block info: %v", err)
+			} else if !blockInfo.Exists {
+				t.Fatalf("TestBoundedMergeDepth: Failed to get block info, block doesn't exists")
+			}
+			return blockInfo.BlockStatus
+		}
+
+		factory := NewFactory()
+		consensusBuild, teardownFunc1, err := factory.NewTestConsensus(params, "BoundedMergeTestBuild")
+		if err != nil {
+			t.Fatalf("TestBoundedMergeDepth: Error setting up consensus: %+v", err)
+		}
+
+		consensusReal, teardownFunc2, err := factory.NewTestConsensus(params, "BoundedMergeTestReal")
+		if err != nil {
+			t.Fatalf("TestBoundedMergeDepth: Error setting up consensus: %+v", err)
+		}
+		defer teardownFunc2()
+
+		// Create a block on top on genesis
+		block1 := buildAndInsertBlock(consensusBuild, []*externalapi.DomainHash{params.GenesisHash})
+
+		// Create a chain
+		selectedChain := make([]*externalapi.DomainBlock, 0, finalityInterval+1)
+		parent := consensusserialization.BlockHash(block1)
+		// Make sure this is always bigger than `blocksChain2` so it will stay the selected chain
+		for i := 0; i < finalityInterval+2; i++ {
+			block := buildAndInsertBlock(consensusBuild, []*externalapi.DomainHash{parent})
+			selectedChain = append(selectedChain, block)
+			parent = consensusserialization.BlockHash(block)
+		}
+
+		// Create another chain
+		blocksChain2 := make([]*externalapi.DomainBlock, 0, finalityInterval+1)
+		parent = consensusserialization.BlockHash(block1)
+		for i := 0; i < finalityInterval+1; i++ {
+			block := buildAndInsertBlock(consensusBuild, []*externalapi.DomainHash{parent})
+			blocksChain2 = append(blocksChain2, block)
+			parent = consensusserialization.BlockHash(block)
+		}
+
+		// Teardown and assign nil to make sure we use the right DAG from here on.
+		teardownFunc1()
+		consensusBuild = nil
+
+		// Now test against the real DAG
+		// submit block1
+		processBlock(consensusReal, block1, "block1")
+
+		// submit chain1
+		for i, block := range selectedChain {
+			processBlock(consensusReal, block, fmt.Sprintf("selectedChain block No %d", i))
+		}
+
+		// submit chain2
+		for i, block := range blocksChain2 {
+			processBlock(consensusReal, block, fmt.Sprintf("blocksChain2 block No %d", i))
+		}
+
+		// submit a block pointing at tip(chain1) and on block2
+		mergeDepthViolatingBlockBottom, isViolatingMergeDepth := checkViolatingMergeDepth(consensusReal, []*externalapi.DomainHash{consensusserialization.BlockHash(blocksChain2[0]), consensusserialization.BlockHash(selectedChain[len(selectedChain)-1])})
+		if !isViolatingMergeDepth {
+			t.Fatalf("TestBoundedMergeDepth: Expected mergeDepthViolatingBlockBottom to violate merge depth")
+		}
+
+		mergeDepthViolatingTop, isViolatingMergeDepth := checkViolatingMergeDepth(consensusReal, []*externalapi.DomainHash{consensusserialization.BlockHash(blocksChain2[len(blocksChain2)-1]), consensusserialization.BlockHash(selectedChain[len(selectedChain)-1])})
```

Comment:
Please add a comment explaining what is being checked here.

Reply:
You assume I remember ;)